### PR TITLE
feat(code/autofix): add preliminary backend support for autofix [ML-62]

### PR DESCRIPTION
**Pull request by agronskiy**
### Description

- Added the structures modelling autofix requests and converters
- Added the backend service parts 
- Added simple tests
- Checked that integrations work

### Checklist

- [x] Tests added and all succeed
- [x] Linted
- [ ] ~~README.md updated, if user-facing~~
- [ ] ~~License file updated, if new 3rd-party dependency is introduced~~

### Screenshots / GIFs

_Visuals that may help the reviewer. Please add screenshots for any UI change. GIFs are most welcome!_

### CODE REVIEW

File: infrastructure/snyk_api/fake_api_client.go
Comment (agronskiy) [reply infrastructure/snyk_api/fake_api_client.go:44]: 😅

File: domain/snyk/command.go
Comment (agronskiy) [reply domain/snyk/command.go]: 🦅 👁️ ! Removed!

File: application/config/config.go
Comment (bastiandoetsch) [inline application/config/config.go:164]: Most likely, we should have a capability to have scanners/infrastructure implementations provide their own configuration and merge it automatically to the top level.

File: infrastructure/code/bundle.go
Comment (bastiandoetsch) [inline infrastructure/code/bundle.go:194]: this could use `compleStatus` I guess.

File: application/config/config.go
Comment (agronskiy) [reply application/config/config.go:164]: Yeah that sounds like a neat idea in a longer term indeed. I guess not only autofix is the candidate to go, but the whole bunch of settings e.g. from Code.

File: infrastructure/code/bundle.go
Comment (agronskiy) [reply infrastructure/code/bundle.go]: ✅

**Comment by agronskiy** [comment]
Calling autofix works on top of the resolves now.

File: application/config/config.go
Comment (agronskiy) [reply application/config/config.go:166]: That would be outside of the scope of this PR, right?

Comment (michelkaporin) [reply application/config/config.go:166]: My point is that it shouldn't be configurable in `config.go`, instead within Code scanner itself, if you want to have it under environment variable.

Comment (michelkaporin) [reply application/config/config.go:166]: Application config should be agnostic to scanner-level configuration, unless it's being exposed as a key to be configured on LS startup / or via user settings in IDE.

Comment (agronskiy) [reply application/config/config.go]: Ok, I moved the settings into `code` package as a singletone too, and added tests. Commit: [fe494b0be564f88cbfc03954cac2c0ebf894df66 ](https://github.com/snyk/snyk-ls/pull/234/commits/fe494b0be564f88cbfc03954cac2c0ebf894df66)

Rejected options:
- moving config as a property of the scanner was less preferrable because the codeaction creation happens at the level of bundle, which has no reverse access to the scanner that created this bundle

Questions about tests:
- Currently, I kept the code for resetting settings defaults in `infrastructure/code` tests only in the respective places (where I test them(, but I would generally like some way to do resetting uniformly. I could import `code` from `testutils` and add `resetCodeSettings` in to the `t.Cleanup` in `testutils.UnitTest(t)` and Co. but that's subject to your decision -- because it makes top-level package aware of the bottom-level ones. Not sure you want that :D

File: infrastructure/code/backend_service.go
Comment (michelkaporin) [inline infrastructure/code/backend_service.go:445]: The context here is the same as the `AnalysisContext` for `AnalysisRequest` (see `analysisRequestBody`). We can extract this to a func.

File: infrastructure/code/backend_service.go
Comment (michelkaporin) [inline infrastructure/code/backend_service.go:424]: You should be careful to depend on anything from `snyk.Issue`. Since it's a domain object, any change to it will impact the autofix behaviour. For instance, `AdditionalData` was meant as a view model for UI so far only. If we decide to change our UI, ruleId dependent logic may break here by mistake. 

As a solution I'd suggest to have own mapping between `snyk.Issue` and the Code issue model that stores its own representation that's needed for autofix. E.g. for ruleId, line number, as they come from `options.issue`.

It's more or less ok for now, but you/we should watch out for regressions here.

File: infrastructure/code/bundle.go
Comment (michelkaporin) [inline infrastructure/code/bundle.go:164]: Still WIP?

File: infrastructure/code/settings.go
Comment (michelkaporin) [inline infrastructure/code/settings.go:37]: Sorry for being grammar nazi - `Singleton`. Singletone sounds either Italian or French, not sure yet 😄

Comment (michelkaporin) [inline infrastructure/code/settings.go:37]: Any reason to have this on a "global" package level? Why not to make it part of the code scanner singleton?

File: infrastructure/code/types.go
Comment (michelkaporin) [inline infrastructure/code/types.go:78]: Isn't it a problem/confusing that it's 1-based on backend, if e.g. analysis results are 0-based on the Code Analysis backend?

File: infrastructure/code/convert.go
Comment (michelkaporin) [inline infrastructure/code/convert.go:38]: This isn't used any more, right?

**Review by michelkaporin — CHANGES_REQUESTED** [review]

File: application/config/config.go
Comment (agronskiy) [reply application/config/config.go:459]: ✅

File: domain/snyk/autofix_suggestion.go
Comment (agronskiy) [reply domain/snyk/autofix_suggestion.go:20]: Very good catch, thanks, moved there

File: infrastructure/code/backend_service.go
Comment (agronskiy) [reply infrastructure/code/backend_service.go:387]: ✅ 🤦

File: infrastructure/code/types.go
Comment (agronskiy) [reply infrastructure/code/types.go:78]: I see your point, actually I was hesitating back and forth if to implement your comment. I confirm that the VScode and LSP are using 0-based ones. I double checked this
> analysis results are 0-based on the Code Analysis backend
and found that it is still 1-based in our API https://github.com/snyk/snyk-ls/blob/7667009db2846f270ab1b4fac2278133fc408836/infrastructure/code/convert.go#L283

This seems to follow the SARIF standard (https://docs.oasis-open.org/sarif/sarif/v2.1.0/sarif-v2.1.0.html#_Ref493492556) which says
> The line number of the first line in a text artifact SHALL be 1

So here I vote to keep conistent with the backend API, I'll add links to comments and make convertation more prominent.

File: infrastructure/code/settings.go
Comment (agronskiy) [reply infrastructure/code/settings.go:37]: ✅

File: infrastructure/code/bundle.go
Comment (agronskiy) [reply infrastructure/code/bundle.go:164]: ✅

File: infrastructure/code/backend_service.go
Comment (agronskiy) [reply infrastructure/code/backend_service.go:424]: https://github.com/snyk/snyk-ls/pull/234#discussion_r1131187779 
> moving config as a property of the scanner was less preferrable because the codeaction creation happens at the level of bundle, which has no reverse access to the scanner that created this bundle

maybe I'm wrong here, you mention that scanner is the singleton but I AFAIU accesing it goes via `snyk` package, then going to its product scanners, filtering them to find the code one etc -- just to arrive back to the code.

As opposed to that chain, keeping it just a config singleton at a package level allows to access it now (and in future) without overhead. Do I oversimplify or miss smth?

File: infrastructure/code/settings.go
Comment (agronskiy) [reply infrastructure/code/settings.go:37]: https://github.com/snyk/snyk-ls/pull/234#discussion_r1131187779 
> moving config as a property of the scanner was less preferrable because the codeaction creation happens at the level of bundle, which has no reverse access to the scanner that created this bundle

maybe I'm wrong here, you mention that scanner is the singleton but I AFAIU accesing it goes via `snyk` package, then going to its product scanners, filtering them to find the code one etc -- just to arrive back to the code.

As opposed to that chain, keeping it just a config singleton at a package level allows to access it now (and in future) without overhead. Do I oversimplify or miss smth?

File: infrastructure/code/backend_service.go
Comment (agronskiy) [reply infrastructure/code/backend_service.go:445]: ✅

File: infrastructure/code/backend_service.go
Comment (agronskiy) [reply infrastructure/code/backend_service.go:424]: I agree with the direction, although detaching the information stored in `Issue` from sounds like not a fully polished solution -- because it opens path to maintaining this map in sync with which issues are not dead (went out of scope) etc. 

You anyways already started with `any` there, so what I would do is (similar to scanners) to have an array of `[]any` in it, and each backend is then free to add its own struct:
```go
import "fmt"

type Issue struct {
	additionalData []any
}

type codeData struct {
	myData string
}

type ossData struct {
	myData int
}

func main() {
	issue := Issue{
		additionalData: []any{},
	}

	// code part creating issue
	issue.additionalData = append(issue.additionalData, codeData{myData: "code"})
	// oss part
	issue.additionalData = append(issue.additionalData, ossData{myData: 42})

	// part consuming the issue
	for _, data := range issue.additionalData {
		codeData, ok := data.(codeData)
		if !ok {
			continue
		}
		fmt.Print("Code data ", codeData.myData)
	}
}
```

All in all, this seems more conformant to the story of having product-specific things on top level (like scanners).

File: infrastructure/code/bundle.go
Comment (agronskiy) [reply infrastructure/code/bundle.go]: ✅

File: infrastructure/code/types.go
Comment (agronskiy) [reply infrastructure/code/types.go]: ✅

File: infrastructure/code/convert.go
Comment (agronskiy) [reply infrastructure/code/convert.go]: ✅

File: infrastructure/code/types.go
Comment (bastiandoetsch) [reply infrastructure/code/types.go]: It's a mess: two standards with different bases. Sigh.

File: infrastructure/code/backend_service.go
Comment (bastiandoetsch) [inline infrastructure/code/backend_service.go:369]: maybe the comment should mention autofix :)

File: infrastructure/code/backend_service.go
Comment (bastiandoetsch) [inline infrastructure/code/backend_service.go:395]: just go over the comments once more - will be helpful in error analysis. Not blocking approval though.

**Review by bastiandoetsch — APPROVED** [review]

File: infrastructure/code/autofix_suggestion.go
Comment (michelkaporin) [inline infrastructure/code/autofix_suggestion.go:24]: Should this belong to `types.go`?

File: infrastructure/code/types.go
Comment (michelkaporin) [reply infrastructure/code/types.go]: Nice find, @agronskiy 👍

**Review by michelkaporin — COMMENTED** [review]
Left few more comms, can't wait it to be merged ❤️ 

How about we update CODEOWNERS file here as well? Or would you rather trust us in making changes to this without your approvals for now? Perhaps if you extract autofix logic to completely separate files, that'd be easier to mark ownership. Hammerhead could still continue owning the basic Code API wrapper implementation.

cc @snyk/hammerhead @snyk/zenith

File: infrastructure/code/fake_snyk_code_api_service.go
Comment (agronskiy) [reply infrastructure/code/fake_snyk_code_api_service.go:122]: ✅

File: infrastructure/code/sarif_types.go
Comment (agronskiy) [reply infrastructure/code/sarif_types.go]: Moved to sarif_utils.go

File: infrastructure/code/autofix_suggestion.go
Comment (agronskiy) [reply infrastructure/code/autofix_suggestion.go:24]: ✅

File: infrastructure/code/backend_service.go
Comment (agronskiy) [reply infrastructure/code/backend_service.go]: ✅

File: infrastructure/code/settings.go
Comment (bastiandoetsch) [inline infrastructure/code/settings.go:98]: shouldn't this be `.js, .java` etc? In this case, I think the regexp must be changed.

File: infrastructure/code/settings.go
Comment (bastiandoetsch) [inline infrastructure/code/settings.go:89]: This is very C-style that you pass in a pointer to a map and then manipulate it. Would be better to return a map with the languages.

File: infrastructure/code/settings.go
Comment (agronskiy) [reply infrastructure/code/settings.go]: @vraychev has an opened PR on sast side which adds additional payload under `/filters` and renders my env var effort unneeded -- I deleted it.

Other than that no, in this case the languages are indeed languages (as in`javascritp`) -- what is stored by the `RunAnalysis` in the `snyk.Issue.AdditionalData` is language model. There's one to many mapping from `language` to `file extenstion`, e.g. engine considers `.ts` as `javascript` for the purposes of analysis. Same for autofix.

File: infrastructure/code/settings.go
Comment (bastiandoetsch) [inline infrastructure/code/settings.go:38]: not a blocker, but maybe better to put into the `codeSettings` type.

File: infrastructure/code/settings.go
Comment (bastiandoetsch) [inline infrastructure/code/settings.go:54]: you probably should use the mutex to avoid races here

File: infrastructure/code/settings.go
Comment (bastiandoetsch) [inline infrastructure/code/settings.go:70]: you probably should use the mutex to avoid races here

**Review by michelkaporin — APPROVED** [review]

**Pull request merged**
